Refactor pipeline related packages into their own plugin modules.

PascalPrecht opened the pull request:
## Overview
One of Embark's goals is to be as extensible as possible. That's why the core strives to be as plugin-driven as possible, meaning that even Embark's core features will be a set of plugins.

This pull request turns all build pipeline related packages into their own plugin modules, and introduce new command handler respectively. 

The command handlers are:

- `pipeline:build` - Performs a build process using webpack and a projects webpack config
- `watcher:start` - Starts a watch process to watch for file changes
- `watcher:stop` - Stops a watch process
- `watcher:restart` - Restarts a watch process

More information can be found in the dedicated commits.

**Please review this carefully** as it touches a core part and APIs of Embark that all of our users use on a daily bases (e.g. `embark run`)

Also, please see the inline comments below.

### Review
@embark-framework/core-devs 

### Cool Spaceship Picture

PascalPrecht commented on lib/modules/watcher/index.js:
```diff
@@ -0,0 +1,180 @@
+
+// TODO: this should be receiving the config object not re-reading the
+// embark.json file
+class Watcher {
```
I renamed this to `Watcher` instead of `Watch`. Thought this was a good opportunity to do this. Let me know if you disagree with this.

PascalPrecht commented:
Travis is unhappy due to linting errors. I'll get those straight once we're done reviewing everything.

PascalPrecht commented on lib/modules/pipeline/index.js:
```diff
@@ -1,64 +1,60 @@
-const utils = require('../utils/utils.js');
-const constants = require('../constants');
+const utils = require('../../utils/utils.js');
+const ProcessLauncher = require('../../core/processes/processLauncher');
```
We probably want to make this one a plugin module as well. Looping in @iurimatias for feedback.

jrainville commented on lib/core/engine.js:
```diff
@@ -219,11 +211,10 @@ class Engine {
-    this.watch = new Watch({logger: this.logger, events: this.events});
-    this.watch.start();
+    this.registerModule('watcher');
+    this.events.request('watcher:start');
```
This looks good

jrainville commented on lib/modules/pipeline/index.js:
```diff
@@ -1,64 +1,60 @@
     this.isFirstBuild = true;
 
-    fs.removeSync(this.buildDir);
+    this.events.setCommandHandler('pipeline:build', (options, callback) => this.build(options, callback))
```
Just an aesthetic suggestion. You can do `this.events.setCommandHandler('pipeline:build', this.build.bind(this))`

PascalPrecht replied:
Yes! I considered that too, but then I my brain found the fat arrow version easier to parse. :)

PascalPrecht commented:
@michaelsbradleyjr thanks for pointing out that the `eject-webpack` command was actually broken. I missed searching inside of `./cmd` for `pipeline` related code, that's how it slipped through.

I've updated the PR accordingly and look into the configuration APIs now.

alaibe commented on lib/core/engine.js:
```diff
@@ -103,21 +103,13 @@ class Engine {
+    this.registerModule('pipeline', {
       webpackConfigName: this.webpackConfigName
     });
+    this.events.emit("status", "Building Assets");
```
should this event being emitted when we actually send:
`self.events.request('pipeline:build'` ?

PascalPrecht replied:
This is an emission I actually haven't introduced in this PR, it only looks like it because I'm moving stuff around. To be honest I don't know what the `status` event is for...

michaelsbradleyjr replied:
`Building Assets` doesn't show up during `embark run` even with `--loglevel trace` (checked on develop branch). Since we already log messages such as `compiling solidity contracts...` and `running webpack with 'development' config...` perhaps we can just remove it? @iurimatias

PascalPrecht commented:
As discussed offline with @michaelsbradleyjr I'll rebase this PR on top of #960 and introduce pipeline config capabilities in the new pipeline module that streamline with other config loading mechanisms within embark.

So let's keep this pending until #960 has landed :)

alaibe commented on lib/modules/watcher/index.js:
```diff
@@ -0,0 +1,180 @@
+        self.events.emit('file-event', 'asset', path);
+      },
+      function () {
+        callback();
```
instead of:
```
function() {
  callback();
}
```

it is possible to replace it with just `callback` ?
(there is the same pattern L96-103

PascalPrecht replied:
Yes, that should be possible. As mentioned above, I haven't changed any code styles in this module (for no particular reason). So this can likely be changed as well :)

alaibe replied:
Feel free to ignore, it appears in the diff that's why I commented but this is mostly code style so minor

alaibe commented on lib/modules/watcher/index.js:
```diff
@@ -0,0 +1,180 @@
+
+    this.watchFiles(
+      filesToWatch,
+      function (eventName, path) {
```
You have used arrow function everywhere except here.
Out of curiosity, any reason for that?

PascalPrecht replied:
No particular reason. I honestly didn't make any code style changes in this module, but I'm happy to change that!

michaelsbradleyjr replied:
Could we add a watcher for a DApp's `webpack.config.js` and `babel-loader-overrides.js` if they exist (no need to watch the default ones inside embark)? Also for the pipeline config?

PascalPrecht replied:
Yes, good point. I can extend this.

PascalPrecht replied:
Okay so I've added a watcher for the `webpack.config.js` inside the dapp, as well as a possible `config/pipeline.js` file. I'm not sure about the `babel-loader-overrides.js` though... Is that something that we expect people to create? Or is that something that gets ejected as well when we do an `embark eject-webpack`?

michaelsbradleyjr replied:
It's something that gets ejected by `embark eject-webpack`.

PascalPrecht commented:
Rebased this and straightened out lint errors. Travis is complaining that `websocket` module isn't available. Not sure if that's part of your work @michaelsbradleyjr ?

PascalPrecht commented on lib/core/config.js:
```diff
@@ -368,10 +372,31 @@ Config.prototype.loadEmbarkConfigFile = function() {
+
+  Object.keys(this.embarkConfig.app).forEach(targetFile => {
+    this.assetFiles[targetFile] = this.loadFiles(this.embarkConfig.app[targetFile]);
+  });
```
Loading asset files has been part of `loadPipelineConfigFile()` before. I just modernized the code a little bit.

What confused me though, was that prior to this commit, `loadPipelineConfigFile()` never loaded a config file, but asset files. So I'm wondering if it makes sense to put that algorithm in another `loadAssetFiles()` functions, or something along those lines that makes more semantic sense. 

Otherwise, now that `loadPipelineConfigFile()` actually **does** load a pipeline config file, we might as well just leave this as is and keep an eye on the fact that it does a little bit more (loading asset files).

Your call.

michaelsbradleyjr replied:
Either way is fine with me, but I would have a slight preference for `loadAssetFiles` being its own function.

michaelsbradleyjr commented on cmd/cmd_controller.js:
```diff
@@ -419,20 +419,16 @@ class EmbarkController {
+    fs.copyPreserve(embarkConfig, dappConfig);
+    console.log(__('webpack config ejected to:').dim.yellow);
     console.log(`${dappConfig}`.green);
+    var embarkOverrides = fs.embarkPath('lib/pipeline/babel-loader-overrides.js');
```
Will need to be changed to `lib/modules/pipeline/babel-loader-overrides.js`, and the `babel-loader-overrides.js` file will need to be moved from `lib/pipeline/` to `lib/modules/pipeline/`.

PascalPrecht replied:
Argh. Good catch! I missed that one

michaelsbradleyjr commented on lib/core/engine.js:
```diff
@@ -103,21 +103,13 @@ class Engine {
-      logger: this.logger,
-      plugins: this.plugins,
+    this.registerModule('pipeline', {
       webpackConfigName: this.webpackConfigName
```
Right, because it's coming from the `--pipeline` cli option/default, and never from a config file as such.

michaelsbradleyjr commented on lib/modules/pipeline/index.js:
```diff
@@ -1,64 +1,60 @@
+    this.events = embark.events;
+    this.logger = embark.config.logger;
+    this.plugins = embark.config.plugins;
     this.webpackConfigName = options.webpackConfigName;
```
Agreed.

michaelsbradleyjr commented on templates/demo/chains.json:
```diff
@@ -1 +1,22 @@
-{}
```
Why does the demo template now have a pre-filled chains.json?

PascalPrecht replied:
This shouldn't be in here. Not sure how this slipped through, but I'll remove this change from this commit.

PascalPrecht commented on lib/modules/watcher/index.js:
```diff
@@ -0,0 +1,201 @@
+
+  watchPipelineConfig(embarkConfig, callback) {
+    const dappPipelineConfigPath = fs.dappPath(embarkConfig.config, DAPP_PIPELINE_CONFIG_FILE);
+    const dappWebpackConfigPath = fs.dappPath(embarkConfig.config, `../${DAPP_WEBPACK_CONFIG_FILE}`);
```
It didn't look like `embarkConfig` exposes the dapp's root, so I have to `../` here .

michaelsbradleyjr replied:
`fs.dappPath` called without arguments will return the dapp's root.

PascalPrecht replied:
Ah, thanks.

PascalPrecht commented on lib/modules/watcher/index.js:
```diff
@@ -0,0 +1,201 @@
+      this.logger.info(`${eventName}: ${path}`);
+      this.events.emit('file-' + eventName, 'config', path);
+      this.events.emit('file-event', 'config', path);
+    }, callback)
```
This is very similar to other `watch*()` methods inside this class. I duplicated this for consistency, but we might want to reconsider introducing multiple watch methods with partially the same algorithm. 

Maybe we can make `Watcher` collect all file paths once and then call `watchFiles()` once for all of them instead of redoing this in several places.

PascalPrecht commented:
Travis again complains about the `websocket` module not being available. I didn't touch any of our dependency configs so I'm unsure how that is possible.

PascalPrecht commented:
Okay I assume this is unrelated to @michaelsbradleyjr changes as #960 is green on Travis, and this PR is rebased on top of that. Will take a deeper look..

PascalPrecht commented:
Turned out one of my commits introduced a change in `npm-shrinkwrap.json`. Resetting that to the state of the previous commit fixed the dependency issue.

Now tests are failing because of how we're handling the retrieval of a config path. 

Working on a fix.

PascalPrecht commented on lib/core/config.js:
```diff
@@ -368,10 +374,33 @@ Config.prototype.loadEmbarkConfigFile = function() {
+  };
+
+  const pipelineConfigPath = fs.dappPath(
+    require(fs.dappPath(EMBARK_CONFIG_FILE)).config || DEFAULT_CONFIG_PATH,
```
This line is failing inside `test_apps/contracts_app`, when running tests, because `fs.dappPath(EMBARK_CONFIG_FILE)).config` is an object as opposed to a path to the config folder inside the dapp.

The reason this is failing on this branch but not in #960 is because #960 has this logic in side `WebpackProcess` which is not actually launched when `npm run embark test` inside `contracts_app` is executed.

I'm not fully aware why and in which scenarios we expect `embarConfig.config` to **not** be a path (or the other way around), but I'll look into how to work around that.

PascalPrecht replied:
After further research and discussion with the team I've learned that "normal" dapps usually configure just a path for Embark to find config files, while "simple" dapps enable control over single config files, in which case they are defined as part of `embarkConfig.config`.

Updated the PR to address this.

PascalPrecht commented:
Here's why the tests are currently failing: https://github.com/embark-framework/embark/pull/962/commits/1a5258830bcd68e25d7f98c045b945f1910c8ce0#r222988505

PascalPrecht commented:
Alright, addressed all comments from @michaelsbradleyjr. I've also noticed that the file watcher didn't pick up file changes from pipeline related config files due to a bug in the path being created.

This has been fixed. I've tested this quadruple times locally, but please take another look at it and let me know if further changes are needed :)

PascalPrecht commented on lib/modules/watcher/index.js:
```diff
@@ -150,6 +162,27 @@ class Watch {
+      filesToWatch.push(embarkConfig.config.pipeline);
+    } else if (typeof embarkConfig.config === 'string') {
+      filesToWatch.push(fs.dappPath(embarkConfig.config, DAPP_PIPELINE_CONFIG_FILE));
+    }
```
Here's another spot where we need to be careful. `config` could be an object ("simple app") but `pipeline` may not be defined because it might be an existing app that doesn't know this config yet and we want to stay backwards compatible.

If it doesn't have a `pipeline` property, we still have to check whether it's a string because it could still be an object, but `fs.dappPath()` will blow up if its parameters aren't strings.
